ENH: Cosine Sine Decomposition

balopat opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Fixes #11429. 

#### What does this implement/fix?
<!--Please explain your changes.-->
Adds wrappers, a convenience function and tests to the scipy.linalg package for Cosine Sine Decomposition. 

#### Additional information
<!--Any additional information you think is important.-->

This is still an early preview, so that I can get early feedback. 
TODO: 

- [x] figure out boolean flags (job flags, transpose, etc.) 
- [x] confirm that my `_lwork` pattern is kosher
- [x] figure out call statement and callprotoargument
- [x] see if we can limit the workspace query inputs or if all is needed 
- [x] decide how to include _construct_csd method as a feature 
- [x] I'm pretty sure this is going to break all sorts of checks for now, fix all those (style, conventions, etc.) 
- [x] wordsmith a proper commit message
- [x] figure out how to give proper attribution to @anuesseler for his pycsd work that saved me a TON of time
- [x] figure out if I need to return RWORK - in case the underlying ZBBCSD doesn't converge "If INFO > 0 on exit, RWORK(2:R) contains the values PHI(1),   ..., PHI(R-1) that, together with THETA(1), ..., THETA(R), define the matrix in intermediate bidiagonal-block form remaining after nonconvergence. INFO specifies the number of nonzero PHI's."  sounds like a useful thing for debugging?

balopat commented:
> I'll make some minor modifications to the wrappers to enable the booleans.
You rock, thank you! 

> Some points,
> we have to follow the numpydoc standards for docstrings (see other decomp files for examples)
> code

Sounds good. 

> the decomp filename should start with an underscore (though we still have legacy files that don't have it `basic.py` etc.),

👍 

> the function should be added to `linalg.__init__.py`

👍 
 
> For the function name, maybe `cossin` is better? cs_decomp doesn't mean much and sounds matlab-y.

Sure, I'm absolutely open to that, and `cossin` makes sense for Cosine Sine decomposition. I would love to have CS Decomposition or CSD in the docs though somewhere to make it a bit easier to find it as I do believe the readers familiar with the papers will try to search based on these terms. 

> Code attribution might not be needed in the end because the wrapper signature will be changed significantly

As I started the whole work based on his work I would consider it a derivative, and it saved me a ton of time. So, if it is not a big problem I would somehow give credit for that. 

> Also we don't need to point to nag pdfs as they are not archived.

Sounds good - there is no clear documentation about the dimensionality of those identity matrices though and what cases they disappear - so for full clarify we will have to merge the content from it to the docs.

ilayn commented:
> I'll make some minor modifications to the wrappers to enable the booleans.

Minor modifications turned out to be a complete rewrite but seems like functionality is stable. The minimum lwork, lrwork values took some time to figure it out.

> As I started the whole work based on his work I would consider it a derivative, and it saved me a ton of time. So, if it is not a big problem I would somehow give credit for that.

We can mention it in the code comments but from the look of it we won't be salvaging anything from it. Hence, unfortunately, there is no relation to our codebase standards so we have to bite the bullet and rewrite most of it anyways. Note that often the burden is the argument checking and raising meaningful errors. The core functionality often takes 25% of the actual code.

> I would love to have CS Decomposition or CSD in the docs though

I agree and people would stumble on it via the docstring search terms. My suggestion is only about the command name. The rest would have all kinds of explanations involving those terms.

> so for full clarify we will have to merge the content from it to the docs.

I think I figured out how it behaves. Let's think about it when we are there.

balopat commented:
> > I'll make some minor modifications to the wrappers to enable the booleans.
> 
> Minor modifications turned out to be a complete rewrite but seems like functionality is stable. The minimum lwork, lrwork values took some time to figure it out.

Thank you for doing that! How did you come to these conclusions? By digging through the Fortran code? 

> > As I started the whole work based on his work I would consider it a derivative, and it saved me a ton of time. So, if it is not a big problem I would somehow give credit for that.
> 
> We can mention it in the code comments but from the look of it we won't be salvaging anything from it. Hence, unfortunately, there is no relation to our codebase standards so we have to bite the bullet and rewrite most of it anyways. Note that often the burden is the argument checking and raising meaningful errors. The core functionality often takes 25% of the actual code.

Fair enough. Let's see where we end up!

> > so for full clarify we will have to merge the content from it to the docs.
> 
> I think I figured out how it behaves. Let's think about it when we are there.
Ok, I saw you added cs11,..cs22 as output args for x11,...,x22 however based on the nag pdf it is only a temporary matrix in the middle of the algo used for bidiagonalization at the beginning - so not sure whether that is correct.


> I think the return args for this function would be the selling point. So depending on how you often use this function, would it make sense to return the full U, VH and CS arrays?

I think that makes sense, especially for the CS matrix. However in my use case I do make use of the u1,u2,v1h,v2h matrices and the theta angles directly. It feels like these two layers both can make sense. Or are you suggesting that the lapack wrapper could return the pieces and the higher level cossin function would return the matrices? I think I'm okay with that!!!

ilayn commented:
> Thank you for doing that! How did you come to these conclusions? By digging through the Fortran code?

Basically yes. Keeping track of minimum required from call to call.

> Ok, I saw you added cs11,..cs22 as output args for x11,...,x22 however based on the nag pdf it is only a temporary matrix in the middle of the algo used for bidiagonalization at the beginning - so not sure whether that is correct.

As happened a few times before, it seems always redundant to return those strange outputs but then a user comes along and says "we use them for such and such but they are not exposed". Then we must break backwards compatibility etc. Hence if LAPACK exposes it we return it anyways as a rule of thumb.

> Or are you suggesting that the lapack wrapper could return the pieces and the higher level cossin function would return the matrices? I think I'm okay with that!!!

Yes, exactly what I meant. The user possibly won't need the pieces but rather need the decomposition itself. Experts can call the wrapper directly anyways.

balopat commented on scipy/linalg/lapack.py:
```diff
@@ -708,6 +708,16 @@
    cunmrz_lwork
    zunmrz_lwork
 
+   cuncsd
```
I hope you meant within the csd methods - I did that, let me know if I missed something - it seems there is no global alphabetic order.

ilayn replied:
alphabetical in the second letter order (the flavors are common anyways)

ilayn replied:
Can I bother you for this part again? The reason why you placed here is because of some previous mistake we had. If you look at line 699, the order is broken from there (I should have caught this before, sorry about that). It would be great if you can fix the ordering for `geqrt` `gemqrt` `tpqrt` and `tpmqrt`.

balopat commented:
Thank you for all the valuable input and explanation! I committed another iteration, I think we are getting closer. I want to cover the error cases in the tests still and also if you notice any more issues let me know.

balopat commented on scipy/linalg/tests/test_decomp_cossin.py:
```diff
@@ -0,0 +1,43 @@
+@pytest.mark.parametrize('m, p, q, dtype, decimals',
+                         [
+                             (3, 2, 1, np.complex128, 12),
+                             (4, 2, 1, np.complex128, 12),
```
Probably we won't need all these cases here and I want to still have tests for the error cases and the sunblock case

ilayn replied:
you can nest parametrizations (see the added `test_orcsd_uncsd_lwork` below) then it cross products all cases. The DTYPES variable is used just because of this purpose. And we are missing the complex64 and float32 variables here.

balopat replied:
Added test cases

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+from .._lib._util import _asarray_validated
+
+
+def cossin(X, partitioning=None):
```
I am a bit undecided on this. Let's take two scenarios; 
 - I have the subarrays and I want to get the decomposition. What I would really like is to slap a dict or list with `cossin([a,b,c,d])` and be done. 
  - I have big matrix and I'd like to just type `cossin(X, 3, 5)` and be done with it without the trouble of iterables. Also for us we just pass the input through `int()` if they are not None so no funky iterable checks.

Thus, I would argue that `cossin(X, p=None, q=None, compute_u=True, compute_vh=True)` signature seems more usable to me than entering this way but it's just me so I'm open for changing my mind.

balopat replied:
I am going to trust your instincts on this if you say this would be more numpy/scipy-esque. 

The one small reason I'd mention why created the single tuple for partitioning is because in the end we are specifying a shape (of the upper-left block), which is not uncommon in numpy/scipy world as an arg (though admittedly annoying with the parentheses), also p without q or vice versa doesn't make sense, they just make sense together - unless maybe `cossin(X, 3)` would mean `cossin(X, 3, 1)`? What do you think?

Regarding `compute_u`, `compute_vh` - I guess those could be interesting to save time/performance, so I'll add them in.

ilayn commented on scipy/linalg/_decomp_cossin.py:
````diff
@@ -0,0 +1,186 @@
+                        0 & 0 & I_{21} & 0 & 0 & 0
+                        \end{pmatrix} \begin{pmatrix} v_1 & 0 \\
+                        0 & v_2\end{pmatrix}^{\dagger}
+    ```
````
These don't belong here I think

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+    The matrix X is a unitary or an orthogonal matrix, partitioned to X11, X12,
+    X21, X22 subblocks and the decomposition satisfies the following equality:
+
+    .. math::
```
I would have appreciated the ascii version as they are often printed in ipython consoles rather than the actual docs. I think ascii was more readable anyways.

balopat replied:
yeah, I was just experimenting to see if this passes and then I left the ``` in 🤦‍♂. (I am working in little 30 minute slices these last couple of days). Also I wasn't sure what's the guidance regarding big formulas - I am glad to see and I agree too that the ascii version is more readable.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+                    x12=x12,
+                    x21=x21,
+                    x22=x22,
+                    compute_u1=True,
```
They are always on?

balopat replied:
yeah, I thought turning off the computation could be considered an "advanced" flow, and those who need it could drop down to the LAPACK level. I have no use case in mind for skipping the computations, but if there are (the mere existence of these flags suggest it) they could save performance, I assume. So I'm okay to plumb `compute_u` and `compute_vh` back to the top.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+    p, q = np.shape(x11)
+    m = np.shape(x22)[0] + p
+
+    iscomp = np.iscomplexobj(x11)
```
You have to check all 4 of them to see if anyone is complex.

balopat replied:
good catch, I was wondering about that while writing...thanks!

balopat replied:
now this led me down a path of finding the largest type and casting all objects to that. 
Let me know what you think. Also wrote a test to cover this case with mixed types.

ilayn replied:
Nice! However `get_lapack_funs` is designed specifically for this. You just need to provide all the arrays `[x11, x12, x21, x22]` as the second argument and it will give the correct flavor and the rest (up/down casting) is done via `f2py`.

balopat replied:
Yes, that's how I started - by `np.any(np.iscomplexobj(...)...)` for the prefix decision - but then, I'll still need the largest type explicitly for the CS construction in https://github.com/scipy/scipy/pull/11524/files#diff-aa0b4763c42b00d6594f8e5df1bd90a3R170

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+    if p != p2 or q != q2 or mmp != mmp2 or mmq != mmq2 or mmq + q != mmp + p:
+        shapes = [np.shape(np.atleast_2d(subx)) for subx in X]
+        raise ValueError("invalid submatrix dimensions {} to build a square "
+                         "matrix, maybe missing partitioning?".format(shapes))
```
`maybe you forgot to use "partitioning" keyword?`

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+    return U, CS, VDH
+
+
+def _construct_cs(theta, m, p, q, dtype):
```
This is a bit too slow with all the for loops, I'll try to cook up something later with less looping.

balopat replied:
I had [a version that was using blocks](https://github.com/scipy/scipy/blob/7aa3e61d7f472a6e459e990c75887a03059d576e/scipy/linalg/tests/test_decomp_cossin.py#L79-L102) do you think that would work?

balopat replied:
The blocks version is not faster at all. I find your solution to be efficient. The cosine-sine loop is required, we have at most 2 identity loops. I might be missing something, but unless there is some kind of numpy magic that copies larger arrays more efficiently into a diagonal than just a simple loop, I think this is optimal.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,186 @@
+    R-by-R nonnegative diagonal matrices satisfying C^2 + S^2 = I, in
+    which R = MIN(P,M-P,Q,M-Q).
+
+    With dimensions
```
We need some markup either code or math for these.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,194 @@
+    The matrix X is a unitary or an orthogonal matrix, partitioned to X11, X12,
+    X21, X22 subblocks and the decomposition satisfies the following equality:
+
+    .. code-block:: none
```
You can remove this rubric here and double the colons at end of the line above `equality::` to mark the following 4 space indented block as ascii code block.

balopat replied:
That's cool, thank you!

balopat commented:
I believe I addressed all your comments. 
A couple of things outstanding: 

- [ ] cossin signature: `cossin(X, p, q)` vs `cossin(X, (p,q))` ? 
- [ ] I still believe there would be value in returning u1,u2,v1h,v2h,theta for the same args as `cossin`. The reason: for my use case I will have to write all this code to handle _lwork, interpret info, etc: 
<details> 

```
U = unitary_group.rvs(M)
p, q = (500,500)

# ----- 
x11 = U[:p, :q]
x12 = U[:p, q:]
x21 = U[p:, :q]
x22 = U[p:, q:]

iscomp = np.any([np.iscomplexobj(sub) for sub in [x11, x12, x21, x22]])
func_name = "uncsd" if iscomp else "orcsd"

csd, csd_lwork = get_lapack_funcs([func_name, func_name + "_lwork"],
                                  [x11, x12, x21, x22])
lwork = scipy.linalg.lapack._compute_lwork(csd_lwork, m=m, p=p, q=q)

lwork_args = (dict(lwork=lwork[0], lrwork=lwork[1]) if iscomp else
              dict(lwork=lwork))

csd_args = dict(x11=x11,
                x12=x12,
                x21=x21,
                x22=x22,
                compute_u1=True,
                compute_u2=True,
                compute_v1t=True,
                compute_v2t=True,
                trans=False,
                signs=False)

csd_args.update(lwork_args)

*_, theta, u1, u2, v1h, v2h, info = csd(
    **csd_args)
... todo error handling...
```

</details>

- [ ] for those who run into convergence issue, we might want to return RWORK, as in case the underlying ZBBCSD doesn't converge "If INFO > 0 on exit, RWORK(2:R) contains the values PHI(1), ..., PHI(R-1) that, together with THETA(1), ..., THETA(R), define the matrix in intermediate bidiagonal-block form remaining after nonconvergence. INFO specifies the number of nonzero PHI's." sounds like a useful thing for debugging? 

- [ ] do we need a tutorial for CSD [here](http://scipy.github.io/devdocs/tutorial/linalg.html)?

ilayn commented:
Let me do some house-cleaning and a tiny bit faster csd construction so that we start on a firm base.

ilayn commented:
OK. I think this can be now worked out. Notice that I rebased some commits. It would be great if you can also follow [this standard](https://docs.scipy.org/doc/numpy-1.9.0/dev/gitwash/development_workflow.html#writing-the-commit-message) for the commit messages. We also miss the small matrix guards such as what happens if empty array comes or 1x1 array comes with p=1, q=0 etc. But this is indeed much closer to our codebase than it was 5 days ago. So thanks again! 

> cossin signature: cossin(X, p, q) vs cossin(X, (p,q)) ?

I am still on the fence about the signature. Moreover we haven't provided the `signs` keyword to the user. 

> I still believe there would be value in returning u1,u2,v1h,v2h,theta for the same args as cossin. The reason: for my use case I will have to write all this code to handle _lwork, interpret info, etc

I feel like I shouldn't have any opinions about this part. You probably can judge better since you used these before in your work.

> for those who run into convergence issue, we might want to return RWORK, as in case the underlying ZBBCSD doesn't converge "If INFO > 0 on exit, RWORK(2:R) contains the values PHI(1), ..., PHI(R-1) that, together with THETA(1), ..., THETA(R), define the matrix in intermediate bidiagonal-block form remaining after nonconvergence. INFO specifies the number of nonzero PHI's." sounds like a useful thing for debugging?

Because we didn't expose ?BBCSD, they wouldn't be able to do much with that information as the failure is fully at the low level between LAPACK gears. I don't think it would be that critical.

> do we need a tutorial for CSD here?

Up to you, but we would really appreciate it if you can provide some real use cases.

balopat commented:
> OK. I think this can be now worked out. Notice that I rebased some commits. 

This looks great, thank you for the cleanup, it's exciting to see it evolving! 


> It would be great if you can also follow [this standard](https://docs.scipy.org/doc/numpy-1.9.0/dev/gitwash/development_workflow.html#writing-the-commit-message) for the commit messages. 

👍 I started doing that. 

> We also miss the small matrix guards such as what happens if empty array comes or 1x1 array comes with p=1, q=0 etc. 

Added in test cases and validations for these cases. 

> But this is indeed much closer to our codebase than it was 5 days ago. So thanks again!

Thank you for being so responsive, and super helpful! 

> > cossin signature: cossin(X, p, q) vs cossin(X, (p,q)) ?
> 
> I am still on the fence about the signature. 

I think your version is growing on me. It's 2 characters less to type! 

> Moreover we haven't provided the `signs` keyword to the user.

Added. 

> > I still believe there would be value in returning u1,u2,v1h,v2h,theta for the same args as cossin. The reason: for my use case I will have to write all this code to handle _lwork, interpret info, etc
> 
> I feel like I shouldn't have any opinions about this part. You probably can judge better since you used these before in your work.

OK. I'll think about how we could introduce this. 
What do you think about a flag (e.g. `low_level: bool = False`) that would return the 5 parameters instead of the 3 matrices? 

> > for those who run into convergence issue, we might want to return RWORK, as in case the underlying ZBBCSD doesn't converge "If INFO > 0 on exit, RWORK(2:R) contains the values PHI(1), ..., PHI(R-1) that, together with THETA(1), ..., THETA(R), define the matrix in intermediate bidiagonal-block form remaining after nonconvergence. INFO specifies the number of nonzero PHI's." sounds like a useful thing for debugging?
> 
> Because we didn't expose ?BBCSD, they wouldn't be able to do much with that information as the failure is fully at the low level between LAPACK gears. I don't think it would be that critical.

Sounds good! 

> > do we need a tutorial for CSD here?
> 
> Up to you, but we would really appreciate it if you can provide some real use cases.

OK, I'll think about it - this might fall into a follow-up PR, just to get the functionality first out of the door. 

> EDIT: I added some ascii box drawing characters but it tripped the unicode checker, when working with the `_decomp_cossin.py`, please add `# -*- coding: utf-8 -*-` all the way at the top as the first line to change the encoding.

Done.

balopat commented:
Looking at the current CI test failures, they seem interesting: 

```
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw1] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw1' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[True-4-1-2-float32]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw0] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw0' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[True-4-1-2-complex64]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw2] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw2' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[True-4-1-2-float64]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw4] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw4' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[False-4-1-2-float64]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw6] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw6' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[False-4-1-2-float32]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw3] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw3' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[True-4-1-2-complex128]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw7] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw7' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[False-4-1-2-complex128]'
______________________ linalg/tests/test_decomp_cossin.py ______________________
[gw8] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
worker 'gw8' crashed while running 'linalg/tests/test_decomp_cossin.py::test_cossin[False-4-1-2-complex64]'
```

This does not happen on Mac. 
Locally on my Linux workstation pytest just gives up on the 4-1-2 case! 

```
python runtests.py -t scipy/linalg/tests/test_decomp_cossin.py -v
Building, see build.log...
Build OK (0:00:02.503915 elapsed)
========================================================== test session starts ===========================================================
platform linux -- Python 3.7.5rc1, pytest-5.3.5, py-1.8.1, pluggy-0.13.1 -- /usr/local/google/home/balintp/.envs/scipy37/bin/python
cachedir: .pytest_cache
rootdir: /usr/local/google/home/balintp/dev/proj/scipy, inifile: pytest.ini
collected 87 items                                                                                                                       

scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-2-1-1-float32] PASSED                                                   [  1%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-2-1-1-float64] PASSED                                                   [  2%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-2-1-1-complex64] PASSED                                                 [  3%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-2-1-1-complex128] PASSED                                                [  4%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-2-1-float32] PASSED                                                   [  5%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-2-1-float64] PASSED                                                   [  6%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-2-1-complex64] PASSED                                                 [  8%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-2-1-complex128] PASSED                                                [  9%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-1-2-float32] PASSED                                                   [ 10%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-1-2-float64] PASSED                                                   [ 11%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-1-2-complex64] PASSED                                                 [ 12%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-3-1-2-complex128] PASSED                                                [ 13%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-4-2-2-float32] PASSED                                                   [ 14%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-4-2-2-float64] PASSED                                                   [ 16%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-4-2-2-complex64] PASSED                                                 [ 17%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-4-2-2-complex128] PASSED                                                [ 18%]
scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-4-1-2-float32] ⏎                                                              
```
and returns with 0 exit code!

balopat commented:
Ah, it is an error handling problem that slips through with one of the arguments: 

```
python -c "from scipy.linalg import cossin;from scipy.stats import unitary_group; cossin(unitary_group.rvs(4), 1, 2)"
 ** On entry to ZUNCSD parameter number 22 had an illegal value
```

balopat commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -92,73 +83,63 @@ def cossin(X, partitioning=None, compute_u=True, compute_vh=True):
+    csd, csd_lwork = get_lapack_funcs([driver, driver + "_lwork"],
+                                      [x11, x12, x21, x22])
+
+    lwork = _compute_lwork(csd_lwork, m=p+q, p=p, q=q)
```
`m=p+q` is incorrect

balopat replied:
> Ah, it is an error handling problem that slips through with one of the arguments:
> 
> ```
> python -c "from scipy.linalg import cossin;from scipy.stats import unitary_group; cossin(unitary_group.rvs(4), 1, 2)"
>  ** On entry to ZUNCSD parameter number 22 had an illegal value
> ```

@ilayn - this worries me that this error slipped through - shouldn't this be handled by the get_lapack_funcs / lapack wrappers and return a nice error status with `info=-22`? Instead it just crashed seemingly!

balopat replied:
yeah, I confirmed that starting with `python runtests.py --ipython` behaves differently on Linux than on Windows.
By setting  L130 (incorrectly) to    `lwork = _compute_lwork(csd_lwork, m=3, p=p, q=q)` and  in ipython: 

```
 from scipy.linalg import cossin; from scipy.stats import unitary_group; cossin(unitary_group.rvs(4), 1, 2) 
```

On MacOSX: (4, 1, 2) passes!! However `cossin(unitary_group.rvs(4), 2, 2)` fails with 
```
ry_group.rvs(4), 2, 2) 

~/dev/proj/scipy/build/testenv/lib/python3.7/site-packages/scipy/linalg/_decomp_cossin.py in cossin(X, p, q, minus_upper, compute_u, compute_vh)
    159     if info < 0:
    160         raise ValueError('illegal value in argument {} of internal {}'
--> 161                          .format(-info, method_name))
    162     if info > 0:
    163         raise LinAlgError("{} did not converge: {}".format(method_name, info))

ValueError: illegal value in argument 24 of internal zuncsd
``` which is good I suppose. 

On Linux (in a docker container): 

```
 ** On entry to ZUNCSD parameter number 22 had an illegal value
```
and the ipython process crashes!

balopat replied:
I guess the difference in parameter number is due to difference in how ILAENV returns different sizes for parameters on different OSes. And for some reason the lapack errors crash scipy on linux and these crashes can't be caught. So...I am guessing that we can't really do much about this.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,199 @@
+    q : int
+        Number of columns of the upper left block ``X11``
+    minus_upper : bool, optional
+        if `False`, the -S, -I block will be the bottom left, otherwise (by
```
Single backticks are for linking to other functions. Double are for printing variables. It's strange but that's why I doubled at the start but forgot these.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,199 @@
+        Number of rows of the upper left block ``X11``
+    q : int
+        Number of columns of the upper left block ``X11``
+    minus_upper : bool, optional
```
I thought this is done via lapack itself via the signs keyword. Also it seems to me that swap_sign or something like that would be a better candidate for the name.

balopat replied:
Yeah, the signs keyword takes care of the `U` and `VH` matrices but `theta` stays the same - how we have to take this into consideration ourselves when we construct the `CS` matrix.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,199 @@
+    n22 = min(m - p, m - q) - r
+    Id = np.eye(np.max([n11, n12, n21, n22, r]), dtype=theta.dtype)
+    CS = np.zeros((m, m), dtype=theta.dtype)
+
```
This will upcast the dtypes for float32 and complex 64. Try `type(-1*np.complex64(1.))`. Better to keep it as float32 as it will be upcasted only when needed.

ilayn replied:
Or use

```
... = -s if swap_sign else s
```

To eliminate sgn completely.

balopat commented:
> Almost ready 😀. I'll be out for a week for holidays but others would pick it up.
Sounds good, enjoy! :) 

> Regarding variable number of return variables, we are really avoiding that as much as we can.
I will then consider doing that in a separate PR and a separate function.

ilayn commented:
I don't know if lowlevel is a good candidate for public api. We need to find a more clever solution such that the number of outputs don't change but if needed more info is returned.

For example, let's say we introduce a keyword `separate` such that when true returmed arguments are tuples of u1 u2 and v1h v2h. And cs is basically theta. Don't take the names too seriously please. Just a proposition. But the idea is that we want as much info packed into as little moving parts as possible without overdoing it. Otherwise we might end up like matlab where things hardly ever make sense in terms of UX. 



I'm literally sitting on a surf board now.

balopat commented:
> I don't know if lowlevel is a good candidate for public api. We need to find a more clever solution such that the number of outputs don't change but if needed more info is returned.

Yeah, I don't love it either. 

> For example, let's say we introduce a keyword `separate` such that when true returned arguments are tuples of u1 u2 and v1h v2h. And cs is basically theta. Don't take the names too seriously please. Just a proposition. But the idea is that we want as much info packed into as little moving parts as possible without overdoing it. Otherwise we might end up like matlab where things hardly ever make sense in terms of UX.

I like the direction of squeezing both levels in to the same function. However using tuples just solves the problem of "number of returned items" but breaks the type of returned items...that still irks me out a bit! What if we return a typed object, e.g a `DecompositionResult` - that could have the individual pieces and the methods to return the matrices? 

> I'm literally sitting on a surf board now. So yes very relaxed proposition 😊

Hahaha, that's awesome, I'm jealous :D 
I also love surfing!

ilayn commented:
> However using tuples just solves the problem of "number of returned items" but breaks the type of returned items...that still irks me a bit!

Welcome to Python-land 🙃

Joking aside. That's more natural than having two competing public funcs is really bad. Also user knows how to slice u and v since they decide the p and q. Then the remaining part is returning theta instead of cs which is same type but different shape. So I'm kind of leaving the slicing to the user and only provide theta - cs switch with some keyword.

balopat commented:
> > However using tuples just solves the problem of "number of returned items" but breaks the type of returned items...that still irks me a bit!
> 
> Welcome to Python-land 🙃
> 
> Joking aside. That's more natural than having two competing public funcs is really bad. Also user knows how to slice u and v since they decide the p and q. Then the remaining part is returning theta instead of cs which is same type but different shape. So I'm kind of leaving the slicing to the user and only provide theta - cs switch with some keyword.

Sounds good :) I added a (tuple, theta, tuple) vs (U, CS, V) version. Let me know if you like it. I like this because then it keeps the "level of abstraction" consistent between the two options (instead of mixing them with (u1, u2, CS, v1, v2)).

balopat commented:
Friendly ping @ilayn, do you think this can be merged soon?

ilayn commented:
Yes! Thanks still ramping up the engine after holidays. 

@mdhaber @larsoner and others, care to go through this ? Especially the theta, cs switch is still undecided for me a bit.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    ``min(p, m - q) - r``, ``min(m - p, q) - r``, and ``min(m - p, m - q) - r``
+    respectively.
+
+    X can be supplied either by itself or by its subblocks in an iterable. See
```
Maybe a bit more explicit text here? 

```
X can be supplied either by itself and block specifications p, q or its subblocks in an iterable from which the shapes would be derived
```
```

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+        of subblocks ``X11``, ``X12``, ``X21``, ``X22``, when ``p``, ``q`` are
+        omitted.
+    p : int
+        Number of rows of the upper left block ``X11``
```
`Used only when X is given as an array`

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    p : int
+        Number of rows of the upper left block ``X11``
+    q : int
+        Number of columns of the upper left block ``X11``
```
`Used only when X is given as an array`

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    Returns
+    -------
+    u : ndarray
+        When `compute_u=True`, contains the block diagonal orthogonal/unitary
```
Double backquotes

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+        The cosine-sine factor with the structure described above.
+         If `separate=True`, this contains the ``theta`` array.
+    vh : ndarray
+        When `compute_vh=True`, contains the block diagonal orthogonal/unitary
```
double backquotes

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    -------
+    u : ndarray
+        When `compute_u=True`, contains the block diagonal orthogonal/unitary
+        matrix consisting of the blocks `U1`` ``(p,p)`` and ``U2`` ``(m-p,m-p)``
```
missing a quote before U1

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    u : ndarray
+        When `compute_u=True`, contains the block diagonal orthogonal/unitary
+        matrix consisting of the blocks `U1`` ``(p,p)`` and ``U2`` ``(m-p,m-p)``
+        orthogonal/unitary matrices. If `separate=True`, this contains the tuple
```
double backquotes

ilayn replied:
`separate=True` is still missing a set of backticks :)

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    vh : ndarray
+        When `compute_vh=True`, contains the block diagonal orthogonal/unitary
+        matrix consisting of the blocks ``V1H`` ``(q,q)`` and
+        ``V2H`` ``(m-q,m-q)`` orthogonal/unitary matrices. If `separate=True`,
```
double around separate=True

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    >>> np.allclose(x, u @ cs @ vdh)
+    True
+
+    Same can be entered via subblocks without the need of ``p``and ``q``. Also
```
One whitespace character after ``p``

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    CS[n11 + r:
+       n11 + r + n12,
+    n11 + n21 + n22 + 2 * r:
+    n11 + n12 + n21 + n22 + 2 * r] = (- Id[:n12, :n12] if not swap_sign
```
You can swap the arguments of if else to avoid the extra `not` clause. A bit easier to read.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    CS = np.zeros((m, m), dtype=theta.dtype)
+
+    CS[:n11, :n11] = Id[:n11, :n11]
+    CS[n11 + r:
```
I would say let's tidy this part a bit via 
```
xs, ys = n11 +r, n11 + n21 + n22 + 2 * r
xe, ye = xs + n12, ys + n12
CS[xs:xe, ys:ye] = Id[:n12, :n12] if swap_sign else -Id[:n12, :n12]
```
otherwise it's really difficult to read. 4 variable assignments shouldn't cause any performance problems.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+    n11 + n12 + n21 + n22 + 2 * r] = (- Id[:n12, :n12] if not swap_sign
+                                      else Id[:n12, :n12])
+
+    CS[p + n22 + r:
```
Same as above for the indices.

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,211 @@
+
+    CS[n11:n11 + r, n11:n11 + r] = c
+    CS[p + n22:p + n22 + r, r + n21 + n22:2 * r + n21 + n22] = c
+    CS[n11:n11 + r,
```
Same with this guy

ilayn commented on scipy/linalg/lapack.py:
```diff
@@ -48,6 +48,11 @@
 .. autosummary::
    :toctree: generated/
 
+   sorcsd
```
Ideally these should show up in line 394 and similarly 685 for uncsd but somehow the latest additions didn't respect the alphabetical order apparently ☹️

balopat commented:
> I did one final sweep. Then I think this can be merged.

Awesome! Thank you for the precise review! I addressed all your comments.

balopat commented:
I think we might have a flake here

```
scipy/optimize/tests/test_minpack.py:732: in test_curvefit_covariance
    assert_allclose(pcov1, pcov2, atol=1e-14)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=1e-14
E   
E   Mismatched elements: 1 / 4 (25%)
E   Max absolute difference: 9.5518202e-08
E   Max relative difference: 1.07895289e-07
E    x: array([[0.885286, 0.212682],
E          [0.212682, 0.058248]])
E    y: array([[0.885286, 0.212682],
E          [0.212682, 0.058248]])
=========================== short test summary info ============================
FAILED scipy/optimize/tests/test_minpack.py::TestCurveFit::test_curvefit_covariance
```

WarrenWeckesser commented:
> I think we might have ran into an unrelated flake here

@balopat, yes, that failure is unrelated.  See https://github.com/scipy/scipy/issues/11652

ilayn commented on scipy/linalg/_decomp_cossin.py:
```diff
@@ -0,0 +1,219 @@
+        The cosine-sine factor with the structure described above.
+         If ``separate=True``, this contains the ``theta`` array.
+    vh : ndarray
+        When `compute_vh=True`, contains the block diagonal orthogonal/unitary
```
compute_vh still has single backticks here.

balopat replied:
Thank you!

balopat commented:
@ilayn this should be good to go. 

CI is failing due to the #11652 flake: 

```
________________________________ test_lsmr[COO] ________________________________
[gw2] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/scipy/sparse/linalg/tests/test_pydata_sparse.py:75: in test_lsmr
    assert_allclose(res[0], res0[0], atol=1e-5)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=1e-05
E   
E   Mismatched elements: 4 / 9 (44.4%)
E   Max absolute difference: 1.65523639e-05
E   Max relative difference: 0.00042785
E    x: array([-0.030391,  0.227403,  0.055178,  0.146272, -0.372236,  0.186547,
E          -0.078312,  0.283226, -0.082756])
E    y: array([-0.030404,  0.227412,  0.055174,  0.146287, -0.372228,  0.18654 ,
E          -0.078316,  0.283238, -0.082773])
____________________ TestCurveFit.test_curvefit_covariance _____________________
[gw0] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/scipy/optimize/tests/test_minpack.py:732: in test_curvefit_covariance
    assert_allclose(pcov1, pcov2, atol=1e-14)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=1e-14
E   
E   Mismatched elements: 1 / 4 (25%)
E   Max absolute difference: 9.5518202e-08
E   Max relative difference: 1.07895289e-07
E    x: array([[0.885286, 0.212682],
E          [0.212682, 0.058248]])
E    y: array([[0.885286, 0.212682],
E          [0.212682, 0.058248]])
```

ilayn commented:
OK did the final sweep and rebased. It's my pleasure to hit the green button. Thank you very much for your efforts and great first PR @balopat !

Do you mind adding a few words in https://github.com/scipy/scipy/wiki/*Release-note-entries-for-SciPy-1.5.0* for the release notes? 

Also don't forget to add an entry to https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*

balopat commented:
Thank you so much @ilayn for all your detailed, hard work reviewing my first PR! I'm very excited to use this function! I updated the wiki pages.

balopat commented:
@ilayn do you know by any chance when 1.5.0 is planned to be released? If this is months away - is there any chance to get this PR cherry-picked in to an earlier release?

ilayn commented:
Under normal circumstances, it will be around June (~6 months cycle). Currently there is no plans for any other release in between.

balopat commented:
Thank you for your quick response!!

tylerjereddy commented:
Even if another release did happen in the `1.4.x` series, it would be unusual to include an enhancement rather than just bug fixes.

balopat commented:
Gotcha, thank you @tylerjereddy! 

Any advice around how to integrate with master version of scipy would be welcome: I would like to raise a draft PR of my new functionality that depends on this version of SciPy and have our CI system pass...
Our library does work on Mac / Windows / Linux. 

I did skim through a little bit the release setup for SciPy, but I only scratched the surface and I see that it is pretty involved to get a release out. But maybe I missed something - is there a standard way to get "bleeding edge" version for SciPy or dev versions for a particular future release (1.5.0)? I found on https://github.com/MacPython/scipy-wheels that https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com/ should contain wheels daily/weekly from `master` but it seems a bit more outdated than that and whl-s are not there consistently for all version for all platforms either? 

Thank you!

tylerjereddy commented:
I've been working to make the weekly wheels available again: https://github.com/MacPython/scipy-wheels/pull/69

We're going through a bit of a transformation b/c we need to migrate away from rackspace to anaconda.org, not just for SciPy, but for things that SciPy depends on as well.

balopat commented:
Thanks @tylerjereddy - that sounds promising! 
I managed to manually build a linux wheel, that succeeds at least. 
The mac one works at least on my mac - but on travis, the wheel I built locally is broken with openblas dependencies.
And the windows source building, well...I got to the point where I built a broken openblas in a couple of hours. 

It will be great to test against the official weekly binaries on all 3 platforms, thank you for making this work again!

mdhaber commented:
Sorry @ilayn I didn't see your ping. 
Thanks, @balopat!

balopat commented:
Just a quick update: I did manage to build the wheel on Windows as well, I think we'll be fine with this build for a while!

gwgundersen commented:
While working on https://github.com/scipy/scipy/pull/12582, I noticed the following issue. When running the Scipy tests (`python runtests.py -v`), I see four failing tests from `test_decomp_cossin.py`:

```bash
=================================================== short test summary info ====================================================
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-40-12-20-float32] - AssertionError: 
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-100-50-50-float32] - AssertionError: 
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[False-40-12-20-float32] - AssertionError: 
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[False-100-50-50-float32] - AssertionError: 
============== 4 failed, 29652 passed, 2030 skipped, 10990 deselected, 76 xfailed, 6 xpassed in 441.03s (0:07:21) ==============
```

Strangely, the maintainer I am working with did not observe this on his own machine. This _feels_ like maybe an OS issue since:

1. The unit tests pass for him but not me (he uses macOS; I use Ubuntu 20.04).
1. The same unit tests fail for me when using a newly cloned version of `master`.
1. The code and tests appear to rely on LAPACK.

Here's an example of what I see if drop a `pdb.set_trace` call before a failing test:

```python
>>> np.isclose(u1_2, u1)
array([[False,  True, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False]])
>>> u1_2[0,0]
0.058479194
>>> u1[0,0]
-0.058478586
>>> u1_2[0,2]
0.040936727
>>> u1[0,2]
-0.040936757
```

ilayn commented:
Interesting indeed. Let's take the discussion to your PR, if you don't mind @gwgundersen  I'll have a look and reply over there to keep it unfragmented.

WarrenWeckesser commented:
We're getting occasional failures of the test `test_cossin_separate`; for example: https://github.com/scipy/scipy/pull/13048/checks?check_run_id=1384417205

Note that sometimes if I rerun the job, the tests all pass.

I don't know how long those test results will stay available, so here's a copy of the relevant output from pytest:
```
=================================== FAILURES ===================================
________________________ test_cossin_separate[float64] _________________________
scipy/linalg/tests/test_decomp_cossin.py:150: in test_cossin_separate
    assert_allclose(u1_2, u1, rtol=0., atol=10*np.finfo(dtype_).eps)
E   AssertionError: 
E   Not equal to tolerance rtol=0, atol=2.22045e-15
E   
E   Mismatched elements: 4514 / 6400 (70.5%)
E   Max absolute difference: 0.80714081
E   Max relative difference: 2.
E    x: array([[-0.012286, -0.063705, -0.046882, ...,  0.067121,  0.042775,
E           -0.038814],
E          [ 0.189033,  0.030199,  0.020188, ..., -0.041924,  0.135174,...
E    y: array([[ 0.012286,  0.063705, -0.046882, ...,  0.067121, -0.042775,
E            0.038814],
E          [-0.189033, -0.030199,  0.020188, ..., -0.041924, -0.135174,...
        X          = array([[-0.01224596,  0.11111008, -0.05583939, ...,  0.00863433,
        -0.02529069, -0.05999699],
       [-0.0022407...007,  0.07247924],
       [ 0.03405866, -0.05719033,  0.02516655, ...,  0.02967258,
         0.04382338,  0.01344216]])
        _          = 0
        dlw        = <fortran object>
        drv        = <fortran object>
        dtype_     = <class 'numpy.float64'>
        lwval      = 10780
        lwvals     = {'lwork': 10780}
        m          = 250
        p          = 80
        pfx        = 'or'
        q          = 170
        theta      = array([0.00541383, 0.00856157, 0.02748419, 0.02978311, 0.0718637 ,
       0.07503233, 0.07906807, 0.10008097, 0.115568...54, 0.99574407, 1.01896566, 1.02267398, 1.04836181,
       1.05885172, 1.09333057, 1.09419429, 1.13869658, 1.18214442])
        theta2     = array([0.00541383, 0.00856157, 0.02748419, 0.02978311, 0.0718637 ,
       0.07503233, 0.07906807, 0.10008097, 0.115568...54, 0.99574407, 1.01896566, 1.02267398, 1.04836181,
       1.05885172, 1.09333057, 1.09419429, 1.13869658, 1.18214442])
        u1         = array([[ 0.01228641,  0.06370488, -0.0468819 , ...,  0.06712145,
        -0.04277498,  0.03881356],
       [-0.1890334...168,  0.16848187],
       [-0.04684342,  0.05722923, -0.11347679, ..., -0.03189451,
        -0.1163678 , -0.03369047]])
        u1_2       = array([[-0.01228641, -0.06370488, -0.0468819 , ...,  0.06712145,
         0.04277498, -0.03881356],
       [ 0.1890334...168, -0.16848187],
       [ 0.04684342, -0.05722923, -0.11347679, ..., -0.03189451,
         0.1163678 ,  0.03369047]])
        u2         = array([[ 0.09401129,  0.09900626, -0.07349002, ...,  0.10217532,
        -0.0603292 ,  0.0933257 ],
       [-0.1354097...708, -0.01956046],
       [ 0.01315501, -0.09415009,  0.01847025, ..., -0.10221244,
         0.06842986,  0.55079481]])
        u2_2       = array([[-0.09401129, -0.09900626, -0.07349002, ...,  0.10217532,
        -0.0603292 ,  0.0933257 ],
       [ 0.1354097...708, -0.01956046],
       [-0.01315501,  0.09415009,  0.01847025, ..., -0.10221244,
         0.06842986,  0.55079481]])
        v1t        = array([[ 0.02071111, -0.12575399, -0.03392104, ...,  0.0561229 ,
         0.00561223,  0.10164257],
       [-0.0456795...206,  0.02532954],
       [-0.00360381, -0.02058857,  0.0545874 , ..., -0.0498076 ,
        -0.06530985,  0.11992337]])
        v1t_2      = array([[-0.02071111,  0.12575399,  0.03392104, ..., -0.0561229 ,
        -0.00561223, -0.10164257],
       [ 0.0456795...206,  0.02532954],
       [-0.00360381, -0.02058857,  0.0545874 , ..., -0.0498076 ,
        -0.06530985,  0.11992337]])
        v2t        = array([[-0.18492015, -0.12474194,  0.07637733, ...,  0.0724234 ,
         0.02643485,  0.16669119],
       [ 0.1832701...928,  0.17164494],
       [-0.12861266,  0.02723643,  0.05142215, ..., -0.15568402,
        -0.10115197, -0.03535843]])
        v2t_2      = array([[ 0.18492015,  0.12474194, -0.07637733, ..., -0.0724234 ,
        -0.02643485, -0.16669119],
       [-0.1832701...928, -0.17164494],
       [ 0.12861266, -0.02723643, -0.05142215, ...,  0.15568402,
         0.10115197,  0.03535843]])
=========================== short test summary info ============================
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin_separate[float64]
= 1 failed, 29554 passed, 2061 skipped, 10990 deselected, 77 xfailed, 5 xpassed in 544.98s (0:09:04) =

```
There are only a couple lines of output for each array, so this might be a false conclusion, but it looks like the problem is that some of the columns in `x` and `y` (i.e. `u1_2` and `u1`) have opposite signs.  Is there a source of indeterminacy here?  (I am thinking of a similar issue related to eigenvectors. Even if normalized to have a norm of 1, the eigenvector associated with a simple eigenvalue is not unique; if `v` is an eigenvector, then so is `-v`.)

ilayn commented:
Yeah it's probably the same up-to-negative-sign issue. We might take the abs and compare. Let me check very quickly to remember what that test was doing.

WarrenWeckesser commented:
Thanks @ilayn.  I see that the possible issue of the sign convention is mentioned in https://github.com/scipy/scipy/pull/12582.

If the indeterminacy of the signs of the columns is expected, then I guess we should fix the unit test to account for it.  Taking the absolute values could work, but a more careful way to deal with it might be to flip the signs column by column so that the first nonzero value in a column is positive before doing the comparison with `assert_allclose`.
